### ISSUE TITLE
ResonanceMolSupplier hangs on NO2

### ISSUE THREAD

**Issue by jasondbiggs** [rdkit/rdkit#4884]
I can create a molecule for NO2, and a `ResonanceMolSupplier`, but asking the supplier if it is at the end results in a hang.

```
>>> from rdkit import Chem
>>> mol = Chem.MolFromSmiles('O=[N+][O-]')
>>> supp = Chem.rdchem.ResonanceMolSupplier(mol)
>>> supp.atEnd()
Killed: 9
```
I waited for a while but it does not return anything.  I haven't tested all the flags exhaustively, but I do find that if I use 
```
Chem.ResonanceFlags.ALLOW_INCOMPLETE_OCTETS | Chem.ResonanceFlags.UNCONSTRAINED_CATIONS| Chem.ResonanceFlags.UNCONSTRAINED_ANIONS
``` 
then it works just fine.  

The hang is present in rdkit versions 2021.09 and 2021.03, but was not present in 2019.03 (that's all I have installed at the moment).

**Comment by ptosco** [comment]
@jasondbiggs Many thanks for reporting this, I’ll take a look and see what’s going on.

**Comment by cbouy** [comment]
Hi,

I'm having the same bug with `CN(C)C=[N+]=[N-]`.

Tested with the following versions of RDKit:
- `2020.03.1`, `2020.03.6` --> works
- `2020.09.1`, `2020.09.3` --> works, but `supp.atEnd()` returns `True`
- `2021.03.1`, `2021.09.5` --> hangs forever, unless the resonance flags are set as in the first post.

**Comment by sebastianmdick** [comment]
I've been having some problems using the flags `Chem.ResonanceFlags.ALLOW_INCOMPLETE_OCTETS | Chem.ResonanceFlags.UNCONSTRAINED_CATIONS| Chem.ResonanceFlags.UNCONSTRAINED_ANIONS` throughout, because it is generally too permissive and creates too many resonance structures. 
At the same time, I needed a reliable solution for this because it would bring my entire workflow to a standstill with no way of catching the error. 
What has worked for me is to check the following to conditions:
- Check that total charge of molecule is 0 
- Check that some atoms have non-zero formal charges.
If both conditions are True, I use the flag `Chem.ResonanceFlags.ALLOW_CHARGE_SEPARATION`, which so far has always worked for me. 

It seems that not allowing charge separation but then providing a molecule with separated charges leads to the issue we observed (not sure that makes sense?!).

**Comment by sebastianmdick** [comment]
I was too optimistic, this still fails for `c1ccc[s+]c1`, which requires `Chem.ResonanceFlags.UNCONSTRAINED_CATIONS` to not freeze.

**Comment by bp-kelley** [comment]
Fixed by #5701
